File: scripts/sources.py
"""Python-side reader for sources.json, the Apps Script source manifest.

The mirror of scripts/sources.js, for the Python consumers: the pattern parser
in tests/test_spam_detector.py, scripts/validate.py and scripts/prod_health.py.

concat_source() must agree byte-for-byte with sources.js concatSource(). Apps
Script concatenates every .gs file into one global scope before running
anything, and the Python harness parses that concatenation as a single body of
source; if the two loaders disagreed about content or order, the JS/Python
parity check in Phase 7 would be comparing two different programs.
"""
import json
import os

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MANIFEST_PATH = os.path.join(ROOT, 'sources.json')


def manifest():
    """Parse and validate sources.json.

    Validated rather than trusted: a manifest that silently read as empty would
    make every consumer "succeed" against no source at all — the parser would
    find no patterns and the version checks would have nothing to disagree with.

    :raises RuntimeError: if the manifest is missing, malformed or incoherent.
    """
    try:
        with open(MANIFEST_PATH, encoding='utf-8') as fh:
            m = json.load(fh)
    except FileNotFoundError:
        raise RuntimeError('Cannot find the source manifest at %s' % MANIFEST_PATH)
    except json.JSONDecodeError as e:
        raise RuntimeError('sources.json is not valid JSON — %s' % e)

    sources = m.get('sources')
    if not isinstance(sources, list) or not sources:
        raise RuntimeError('sources.json: "sources" must be a non-empty array')
    if not all(isinstance(s, str) and s for s in sources):
        raise RuntimeError('sources.json: every entry in "sources" must be a filename')
    if len(set(sources)) != len(sources):
        dupes = sorted({s for s in sources if sources.count(s) > 1})
        raise RuntimeError('sources.json: duplicate entries in "sources": %s — '
                           'a file listed twice is concatenated twice, which is a '
                           'duplicate-declaration SyntaxError at load time'
                           % ', '.join(dupes))
    bad = [s for s in sources if s.startswith('/') or '..' in s.split('/')]
    if bad:
        raise RuntimeError('sources.json: entries must be repo-relative paths '
                           'without "..": %s' % ', '.join(bad))

    version_file = m.get('versionFile')
    if not isinstance(version_file, str) or not version_file:
        raise RuntimeError('sources.json: "versionFile" must be a non-empty string')
    if version_file not in sources:
        raise RuntimeError('sources.json: versionFile %r is not listed in "sources"'
                           % version_file)
    return m


def source_names():
    """Manifest-ordered source filenames, repo-relative."""
    return list(manifest()['sources'])


def source_paths():
    """Manifest-ordered absolute paths."""
    return [os.path.join(ROOT, f) for f in manifest()['sources']]


def version_file_path():
    """Absolute path of the file carrying the version markers."""
    return os.path.join(ROOT, manifest()['versionFile'])


def concat_source():
    """The deployed script as Apps Script will see it: all files, in order.

    Joined with a newline so a file that does not end in one cannot weld its
    last line onto the next file's first line.
    """
    parts = []
    for path in source_paths():
        try:
            # newline='' disables universal-newline translation. Without it
            # Python silently rewrites CRLF to LF while Node's readFileSync does
            # not, so the two readers would emit different bytes for the same
            # file and the byte-for-byte contract above would be quietly false.
            # test_patch_version.js asserts the sources are LF-only, so this
            # reads through unchanged in practice; it is here so that if that
            # ever stops being true, the parity test fails loudly instead of the
            # difference being papered over.
            with open(path, encoding='utf-8', newline='') as fh:
                parts.append(fh.read())
        except OSError as e:
            raise RuntimeError('sources.json lists %s, which cannot be read — %s'
                               % (os.path.relpath(path, ROOT), e))
    return '\n'.join(parts)

File: scripts/sources.js
/**
 * Node-side reader for sources.json, the Apps Script source manifest.
 *
 * Every Node consumer — the three vm-based test suites, patch_version.js, the
 * .claspignore hygiene check, and the deploy workflow's syntax lint — loads the
 * script through here, so splitting SpamDetector.gs into several .gs files
 * changes one JSON file and nothing else.
 *
 * concatSource() is the important one. Apps Script concatenates every .gs file
 * in the project into a single global scope before running anything, so the
 * faithful way to test the shipped code in a `vm` is to concatenate it the same
 * way. Loading files into separate vm.Scripts would give each its own scope and
 * quietly break every cross-file reference — a test harness that disagrees with
 * production about something that basic is worse than no harness.
 */
'use strict';
const fs   = require('fs');
const path = require('path');

const ROOT          = path.join(__dirname, '..');
const MANIFEST_PATH = path.join(ROOT, 'sources.json');

function manifest() {
  let raw;
  try {
    raw = fs.readFileSync(MANIFEST_PATH, 'utf8');
  } catch (e) {
    throw new Error('Cannot read the source manifest at ' + MANIFEST_PATH +
                    ' — ' + e.message);
  }

  let m;
  try {
    m = JSON.parse(raw);
  } catch (e) {
    throw new Error('sources.json is not valid JSON — ' + e.message);
  }

  // Validated rather than trusted. A manifest that silently reads as empty
  // would make every consumer "succeed" against no source at all: the vm suites
  // would load nothing and pass, and the syntax lint would lint nothing. Fail
  // loudly instead.
  if (!Array.isArray(m.sources) || m.sources.length === 0) {
    throw new Error('sources.json: "sources" must be a non-empty array');
  }
  if (!m.sources.every(s => typeof s === 'string' && s)) {
    throw new Error('sources.json: every entry in "sources" must be a filename');
  }
  const dupes = m.sources.filter((s, i) => m.sources.indexOf(s) !== i);
  if (dupes.length) {
    throw new Error('sources.json: duplicate entries in "sources": ' +
                    [...new Set(dupes)].join(', ') + ' — a file listed twice is ' +
                    'concatenated twice, which is a duplicate-declaration ' +
                    'SyntaxError at load time');
  }
  const bad = m.sources.filter(s => s.startsWith('/') || s.split('/').includes('..'));
  if (bad.length) {
    throw new Error('sources.json: entries must be repo-relative paths without ' +
                    '"..": ' + bad.join(', '));
  }
  if (typeof m.versionFile !== 'string' || !m.versionFile) {
    throw new Error('sources.json: "versionFile" must be a non-empty string');
  }
  if (m.sources.indexOf(m.versionFile) === -1) {
    throw new Error('sources.json: versionFile ' + JSON.stringify(m.versionFile) +
                    ' is not listed in "sources"');
  }
  return m;
}

/** Manifest-ordered source filenames, repo-relative. @return {string[]} */
function sourceNames() {
  return manifest().sources.slice();
}

/** Manifest-ordered absolute paths. @return {string[]} */
function sourcePaths() {
  return manifest().sources.map(f => path.join(ROOT, f));
}

/** Absolute path of the file carrying the version markers. @return {string} */
function versionFilePath() {
  return path.join(ROOT, manifest().versionFile);
}

/**
 * The deployed script as Apps Script will see it: every source file joined in
 * manifest order.
 *
 * Joined with a newline so a file that does not end in one cannot weld its last
 * line onto the next file's first line.
 *
 * @return {string}
 */
function concatSource() {
  return sourcePaths().map(p => {
    try {
      return fs.readFileSync(p, 'utf8');
    } catch (e) {
      throw new Error('sources.json lists ' + path.relative(ROOT, p) +
                      ', which cannot be read — ' + e.message);
    }
  }).join('\n');
}

module.exports = { sourceNames, versionFilePath, concatSource };
